Add webhook support for search alerts

albertisfu opened the pull request:
This PR adds webhook support for search alerts  as described in #854

The webhook event for search alerts is triggered every time a search alert is sent to a user following the same alert rates.

The webhook event JSON hash for search alerts has the following keys:

**results:** contains all the data related to the Search alert, it has all the same fields as the Search API.
**webhook:** data related to the webhook endpoint
**search_alert:** data related to the search alert that triggered the event is based on the Search Alert API. The only difference is that it doesn't show a URL for the API resource as the Search Alert, instead, it only shows the resource ID, this is because to generate the resource URL the `HyperlinkedModelSerializer` requires to pass an `HTTP request`.

Here's an example of a webhook event for a search alert.

```
{
   "results": [
        {
            "absolute_url": "/opinion/2812209/obergefell-v-hodges/",
            "attorney": "",
            "author_id": null,
            "caseName": "Obergefell v. Hodges",
            "caseNameShort": "Obergefell",
            "citation": [
                "192 L. Ed. 2d 609",
                "135 S. Ct. 2584",
                "2015 U.S. LEXIS 4250",
                "576 U.S. 644"
            ],
            "citeCount": 432,
            "cites": [
                ...
            ],
            "cluster_id": 2812209,
            "court": "Supreme Court of the United States",
            "court_citation_string": "SCOTUS",
            "court_exact": "scotus",
            "court_id": "scotus",
            "dateArgued": null,
            "dateFiled": "2015-06-26T00:53:00-07:00",
            "dateReargued": null,
            "dateReargumentDenied": null,
            "docketNumber": "14-556",
            "docket_id": 2668808,
            "download_url": "http://www.supremecourt.gov/opinions/14pdf/14-556_3204.pdf",
            "id": 2812209,
            "joined_by_ids": null,
            "judge": "",
            "lexisCite": "2015 U.S. LEXIS 4250",
            "local_path": null,
            "neutralCite": null,
            "non_participating_judge_ids": null,
            "pagerank": null,
            "panel_ids": null,
            "per_curiam": null,
            "scdb_id": "2014-070",
            "sibling_ids": [
                2812209
            ],
            "snippet": "      Syllabus\n\n  <mark>OBERGEFELL</mark> ET AL. <mark>v</mark>. <mark>HODGES</mark>, DIRECTOR, OHIO \n\n        DEPARTMENT &hellip; the same court.\n2                       <mark>OBERGEFELL</mark> <mark>v</mark>. <mark>HODGES</mark>\n\n                                  Syllabus&hellip;commitment\nto procreate.\n4                       <mark>OBERGEFELL</mark> <mark>v</mark>. <mark>HODGES</mark>\n\n                                  Syllabus&hellip;\n     JAMES <mark>OBERGEFELL</mark>, ET AL., PETITIONERS\n14–556               <mark>v</mark>.\n       RICHARD <mark>HODGES</mark>, DIRECTOR, OHIO&hellip;specific rights that allow\n2                <mark>OBERGEFELL</mark> <mark>v</mark>. <mark>HODGES</mark>\n\n                     Opinion of the Court",
            "source": "C",
            "status": "Precedential",
            "status_exact": "Precedential",
            "suitNature": "",
            "timestamp": "2022-11-14T15:26:19.603000-08:00",
            "type": "010combined"
        }
    ],
   "webhook":{
      "version":1,
      "event_type":2,
      "date_created":"2022-12-02T23:42:34.894411+00:00",
      "deprecation_date":null
   },
   "search_alert":{
      "id":1,
      "name":"Obergefell v. Hodges",
      "rate":"dly",
      "user":1,
      "query":"?q=Obergefell+v.+Hodges&type=o&order_by=score+desc&stat_Precedential=on&filed_after=06%2F01%2F2015&docket_number=14-556&",
      "secret_key":"MV8yfOSXQvDQauQOCEg0orkgi4hBC4r9yg5Wk3ue",
      "date_created":"2022-12-02T15:42:34.921805-08:00",
      "date_last_hit":"2022-12-02T15:42:35.915737-08:00",
      "date_modified":"2022-12-02T15:42:35.915880-08:00"
   }
}
```
The search alert webhook event is triggered independently for every search alert the user has active for a specific rate. This differs from the email search alert where all the search alerts for a rate are sent in the same message.

An alternative might be to change the JSON hash structure in order to send results for multiple alerts in the same event something like:

```
results:[ 
  "alert"{
  "search_alert":{ ....},
  "search_results":[....]
  },
],
webhook{...}
```

Let me know if you have any comments.

albertisfu commented on cl/alerts/management/commands/cl_send_alerts.py:
```diff
@@ -147,7 +154,11 @@ def run_query(self, alert, rate):
-            main_params = search_utils.build_main_query(cd, facet=False)
+            main_params = search_utils.build_main_query(
+                cd,
+                highlight="text",  # Required to show all field as in Search API
```
This change was also required to show all the fields as in the search API, if we omit this all the fields are highlighted and they're not serialized, so I changed keys in the email templates according to this change.

mlissner commented on cl/alerts/factories.py:
```diff
@@ -23,3 +24,17 @@ class DocketAlertWithParentsFactory(DocketAlertFactory, DocketParentMixin):
     pass
+
+
+class AlertFactory(DjangoModelFactory):
```
Will you also add this to the `make_dev_data` script, please?

mlissner commented on cl/alerts/management/commands/cl_send_alerts.py:
```diff
@@ -216,13 +227,22 @@ def send_emails(self, rate):
+                    user_webhooks = user.webhooks.filter(
+                        event_type=WebhookEventType.SEARCH_ALERT, enabled=True
+                    )
+                    if user_webhooks.exists():
+                        self.send_search_alert_webhook(
+                            results, user_webhooks.first(), search_type, alert
+                        )
```
From the docs:

> If the QuerySet has no ordering defined, then the queryset is automatically ordered by the primary key.

I guess your approach here, using `first()` would thus be mostly fine, because we enforce one enabled webhook per event type on the client side, but I could imagine us using the admin panel to create a second webhook for an event type for a particular client that asked us to.

So I think this might be better (and in the normal case would perform just as well):

```suggestion
                    for user_webhook in user_webhooks:
                        self.send_search_alert_webhook(
                            results, user_webhook, search_type, alert
                        )
```

(

Actually, in the case where a user only has one webhook, this would perform better, because you'd have one query:

    SELECT * FROM webhooks where user = x;

Instead of two:

    SELECT * FROM webhooks where user = x LIMIT 1;
    SELECT * FROM webhooks where user = x;

But this difference is very marginal! I just like to think carefully about our poor database.

)

albertisfu replied:
great! yes, this solves the problem if at any moment a user has more than one webhook endpoint.

mlissner commented on cl/users/api_views.py:
```diff
@@ -130,6 +130,17 @@ def test_webhook(self, request, *args, **kwargs):
                 da_dummy_curl = da_curl_template.render(
                     {"endpoint_url": webhook.url}
                 ).strip()
+            case WebhookEventType.SEARCH_ALERT:
+                da_template = loader.get_template(
+                    "includes/search_alert_webhook_dummy.txt"
+                )
+                da_dummy_content = da_template.render().strip()
+                da_curl_template = loader.get_template(
+                    "includes/search_alert_webhook_dummy_curl.txt"
+                )
+                da_dummy_curl = da_curl_template.render(
+                    {"endpoint_url": webhook.url}
+                ).strip()
```
`da_` in this context stands for Docket Alert, doesn't it? Need some variable clean up here?

albertisfu replied:
Of course, updated to `event_` to be general.

mlissner commented on cl/users/templates/includes/search_alert_webhook_dummy.txt:
```diff
@@ -0,0 +1,145 @@
+            ],
+            "citeCount": 432,
+            "cites": [
+                87116,
```
I think I'd leave the `cites` key blank just to keep it simple for folks. Same in the cURL example.

albertisfu replied:
Updated.

mlissner commented on cl/users/templates/includes/webhooks_htmx/webhooks-test-webhook.html:
```diff
@@ -17,7 +17,7 @@
       </div>
     </div>
-  {% if da_type == webhook.event_type %}
+  {% if  webhook.event_type in da_type %}
```
Seems like we need some variable cleanup here too, right?

albertisfu replied:
Yeah, updated to `event_types`.

mlissner reviewed (changes requested):
Looks good. Some of the variable/function/method names need updating. Like, `send_emails` is no longer accurate and a bunch of variables start with `da_`, which doesn't apply to search alerts. 

Can you do a quick look for any that need tweaks? Otherwise, once you've got the structure cleaned up, as in #2395, I think this looks pretty good to me.

albertisfu commented:
Thank you, I've added these changes.

Also updated the search alert webhook structure to:

```
{
payload: {results: [], alert: {}},
webhook: {}
}
```
Updated tests and webhook event test dummy data, accordingly.

Let me know if there is anything else here.

mlissner reviewed (approved):
Looks good, thanks.
